### PR TITLE
Fix migration from local storage to NFS in KVM

### PR DESCRIPTION
**Pull request by JoaoJandre**
### Description

Using KVM as a hypervisor, when performing a live migration of a VM with a volume, if the volume is on local storage and the migration destination is an NFS storage, an error occurs during the migration. This happens because there is a validation to check whether the destination host has direct access to the source storage of the volume being migrated. However, this validation is not necessary, since the host that will perform the migration is the source host itself. Therefore, the destination host only needs access to the destination storage.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### How Has This Been Tested?

Before this PR, the following live volume migration cases failed in KVM:

| Volume |  Origin |  Destination |
| ------ | ------ | ------ |
| ROOT | Local | NFS1 |
| DATA1 | Local | NFS1 |

| Volume |  Origin |  Destination |
| ------ | ------ | ------ |
| ROOT | Local | NFS1 |

With this PR, the above cases now work. Furthermore, the following cases where validated, to make sure that no regressions were introduced:

| Volume |  Origin |  Destination |
| ------ | ------ | ------ |
| ROOT | NFS1 | NFS2 |
| DATA1 | NFS1 | NFS2 |

| Volume |  Origin |  Destination |
| ------ | ------ | ------ |
| ROOT | NFS1 | Local |
| DATA1 | NFS1 | Local |

| Volume |  Origin |  Destination |
| ------ | ------ | ------ |
| ROOT | NFS1 | ISCSI1 |
| DATA1 | NFS1 | ISCSI1 |

| Volume |  Origin |  Destination |
| ------ | ------ | ------ |
| ROOT | ISCSI1 | NFS1 |
| DATA1 | ISCSI1 | NFS1 |

### CODE REVIEW

**Comment by rohityadavcloud** [comment]
@JoaoJandre is this also applicable fix for 4.19?

**Comment by JoaoJandre** [comment]
> @JoaoJandre is this also applicable fix for 4.19?

@rohityadavcloud yes, I'll rebase with 4.19 and target it

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@JoaoJandre : 
```
09:11:09 [ERROR] /jenkins/workspace/acs-centos7-pkg-builder/dist/rpmbuild/BUILD/cloudstack-4.19.1.0-SNAPSHOT/engine/storage/datamotion/src/test/java/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategyTest.java:26:15: Unused import - org.mockito.Mockito.never. [UnusedImports]
09:11:09 [ERROR] /jenkins/workspace/acs-centos7-pkg-builder/dist/rpmbuild/BUILD/cloudstack-4.19.1.0-SNAPSHOT/engine/storage/datamotion/src/test/java/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategyTest.java:51:8: Unused import - org.mockito.verification.VerificationMode. [UnusedImports]
09:11:09 [ERROR] /jenkins/workspace/acs-centos7-pkg-builder/dist/rpmbuild/BUILD/cloudstack-4.19.1.0-SNAPSHOT/engine/storage/datamotion/src/test/java/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategyTest.java:65:8: Unused import - java.util.LinkedList. [UnusedImports]
```

**Comment by JoaoJandre** [comment]
@blueorangutan package

**Comment by JoaoJandre** [comment]
@sureshanaparti could we run the CI here?

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by JoaoJandre** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@JoaoJandre thanks for this fix, I'm trying to reproduce the issue and bug fix, what APIs are you using the replicate this situation. I'm questioning how do you migrate a volume with shared storage offering to local storage?

**Comment by JoaoJandre** [comment]
> @JoaoJandre thanks for this fix, I'm trying to reproduce the issue and bug fix, what APIs are you using the replicate this situation. I'm questioning how do you migrate a volume with shared storage offering to local storage?

I did all tests using live VM + volume migration on the UI. I was able to migrate volumes from NFS to local and from local to NFS this way. Looking at the API call that the UI made, it was using `migrateVirtualMachineWithVolume`. Something like:
```
http://<mgmtserver>/client/api/?hostid=e8050fa0-c834-47c5-9a1e-089595aa2edc&virtualmachineid=27640463-d916-4e9b-b223-4f3c84fc828a&migrateto[0].volume=31df1fe1-53f2-4ff5-9b86-d75f6f2afe85&migrateto[0].pool=49a3f50d-e70c-4963-8e6a-f70e3a9d295d&command=migrateVirtualMachineWithVolume&response=json
```

**Review by borisstoyanov — COMMENTED** [review]
Thanks @JoaoJandre, we have the option when changing disk offering to migrate to another storage if required, why not fix it there instead?
Screenshot 2024-06-12 at 11 44 52

I think it's more lean approach to execute this operation since in the case you've described you might end up with volume on shared storage while having assigned offering of local storage and vice versa. 
Currently migrating cross different types of storage seems not working when changing the offering ( returns 'Index 0 out of bounds for length 0' ), would you consider moving the fix there?

**Comment by JoaoJandre** [comment]
> Thanks @JoaoJandre, we have the option when changing disk offering to migrate to another storage if required, why not fix it there instead? Screenshot 2024-06-12 at 11 44 52
> 
> I think it's more lean approach to execute this operation since in the case you've described you might end up with volume on shared storage while having assigned offering of local storage and vice versa. Currently migrating cross different types of storage seems not working ( returns 'Index 0 out of bounds for length 0' ), would you consider moving the fix there?


Hey @borisstoyanov, 

I think you misunderstood the PR. The issue is not with the disk offering, but with a validation that makes no sense: the destination host does not need to have access to the original storage, regardless of the disk offering. This PR fixes the current validation to reflect that. 

About ending up with a disk offering in a storage with different type, this is current behavior, as introduced in #2425. It is assumed that when the ROOT admin migrates the volume somewhere, they know what they are doing. Same applies to migrating VMs to hosts that do not have a certain tag, the root admin has this power.

I didn't understand your last point, are you saying that this PR broke migration across different types of storage, or that this is a current issue? if so, could you give me the link to the issue?

**Review by borisstoyanov — APPROVED** [review]
I apologise @JoaoJandre, I thought this was much deeper change, what you suggest is making sense now, and the issues I'm pointing out are not related to your PR. 
LGTM

**Comment by JoaoJandre** [comment]
> I apologise @JoaoJandre, I thought this was much deeper change, what you suggest is making sense now, and the issues I'm pointing out are not related to your PR. LGTM

No problem @borisstoyanov :)

**Comment by rohityadavcloud** [comment]
Re-kicking GH actions

**Comment by rohityadavcloud** [comment]
Is this otherwise good to merge @JoaoJandre @sureshanaparti @DaanHoogland ?
@blueorangutan package

**Comment by JoaoJandre** [comment]
> Is this otherwise good to merge @JoaoJandre @sureshanaparti @DaanHoogland ? @blueorangutan package

@rohityadavcloud Only need one more +1

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java [hunk 1]
```diff
@@ -2385,26 +2385,10 @@ protected void verifyLiveMigrationForKVM(Map<VolumeInfo, DataStore> volumeDataSt
             } else if (storageTypeConsistency != destStoragePoolVO.isManaged()) {
                 throw new CloudRuntimeException("Destination storage pools must be either all managed or all not managed");
             }
-
-            addSourcePoolToPoolsMap(sourcePools, srcStoragePoolVO, destStoragePoolVO);
         }
         verifyDestinationStorage(sourcePools, destHost);
```
Comment (sureshanaparti) [inline engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java:2389]: `sourcePools` is empty here

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java [hunk 2]
```diff
@@ -2385,26 +2385,10 @@ protected void verifyLiveMigrationForKVM(Map<VolumeInfo, DataStore> volumeDataSt
...
     }
 
-    /**
-     * Adds source storage pool to the migration map if the destination pool is not managed and it is NFS.
-     */
-    protected void addSourcePoolToPoolsMap(Map<String, Storage.StoragePoolType> sourcePools, StoragePoolVO srcStoragePoolVO, StoragePoolVO destStoragePoolVO) {
```
[the remark this answers is not in the document]
Comment (JoaoJandre) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java:2397]: @sureshanaparti not anymore

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java (hunk 1 above)
Comment (JoaoJandre) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java]: @sureshanaparti yes, I didn't realize that it was only being populated here, I've removed more dead code that is not used anymore, thanks.

**Review by DaanHoogland — APPROVED** [review]
code looks good and trusting the testing done

**Review by sureshanaparti — APPROVED** [review]
clgtm, didn't test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 kvm-alma9

**Pull request merged**